libs/@blockprotocol/graph/src/stdlib/subgraph/element/entity.ts

```diff
@@ -54,80 +66,94 @@ export const getEntity = (
-          targetVersion = revisionVersion;
-          break;
+        if (!isEntityVertex(vertex)) {
+          throw new Error(
+            `Found non-entity vertex associated with EntityId: ${entityId}`,
+          );
         }
-      }
-    }
-
-    entityRevision =
-      targetVersion !== undefined
-        ? entityRevisions[targetVersion]!.inner
-        : undefined;
+        return vertex.inner;
+      })
+      .find((entity) => {
+        return intervalContainsTimestamp(
+          /*
+           these casts are safe as we check for `targetRevisionInformation === undefined` above and that's only ever
+           defined if `Temporal extends true`
+           */
+          (entity as Entity<true>).metadata.temporalVersioning[
+            (subgraph as Subgraph<true>).temporalAxes.resolved.variable.axis
+          ],
+          targetTime,
+        );
+      });
   }
-
-  return entityRevision;
 };
 
 /**
- * Returns all `Entity`s within the vertices of the subgraph that match a given `EntityId`
+ * Returns all {@link Entity} revisions within the vertices of the subgraph that match a given {@link EntityId}.
  *
- * @param subgraph
- * @param entityId
+ * When querying a subgraph with support for temporal versioning, it optionally constrains the search to a given
+ * {@link NonNullTimeInterval}.
+ *
+ * @param {Subgraph }subgraph
+ * @param {EntityId} entityId
+ * @param {NonNullTimeInterval} [interval]
  */
-export const getEntityRevisionsByEntityId = (
-  subgraph: Subgraph,
+export const getEntityRevisionsByEntityId = <Temporal extends boolean>(
+  subgraph: Subgraph<Temporal>,
   entityId: EntityId,
-): Entity[] => {
+  interval?: Temporal extends true ? NonNullTimeInterval : undefined,
+): Entity<Temporal>[] => {
   const versionObject = subgraph.vertices[entityId];
 
   if (!versionObject) {
     return [];
   }
-  const entityVertices = Object.values(versionObject);
 
-  return entityVertices.filter(isEntityVertex).map((vertex) => {
-    return vertex.inner;
-  });
+  if (interval !== undefined) {
+    return typedEntries(versionObject)
+      .filter(([startTime, _entityVertex]) =>
+        intervalContainsTimestamp(interval, startTime),
+      )
+      .map(([_startTime, entityVertex]) => entityVertex)
+      .filter(isEntityVertex)
+      .map((entityVertex) => entityVertex.inner)
+      .filter((entity) =>
+        /*
+        We want to find all entities which were active _at some point_ in the search interval. This is indicated by an
+        overlap.
+         */
+        intervalOverlapsInterval(
+          interval,
+          /*
+           these casts are safe as we check for `targetRevisionInformation === undefined` above and that's only ever
+           defined if `Temporal extends true`
+           */
+          (entity as Entity<true>).metadata.temporalVersioning[
+            (subgraph as Subgraph<true>).temporalAxes.resolved.variable.axis
+          ],
+        ),
+      );
```

Comment:
Looks like a lot of looping – can we not do something like this?
```suggestion
    const entities = [];
    for (const [_startTime, vertex] of typedEntries(versionObject)) {
      const entity = isEntityVertex(vertex) ? vertex.inner : null;

      if (
        entity &&
        "temporalVersioning" in entity.metadata &&
        intervalOverlapsInterval(
          interval,
          entity.metadata.temporalVersioning[
            subgraph.temporalAxes.resolved.variable.axis
          ],
        )
      ) {
        entities.push(entity);
      }
    }
    return entities;
```

Reply:
I was fairly confused by your comment and after some googling I think I'm about as disappointed in JS as I have ever been. I cannot believe that the functional iterator methods each create a loop. That is insane.

Basically every language I've interacted with (even streams in Java) combines functional iterators that can be combined (for example multiple filters in a row, etc.) so that you are iterating a minimal amount of times and so that the compiler can actually be more confident about how to optimise the code compared to for loops with bound checks and such.

I'm going to search across the code for a variety of these now.

Reply:
>  I cannot believe that every functional iterator method creates a loop. 

And an array 😁 (at least these ones)

Reply:
There are libraries that do what you want and you can write your own with generators, but easiest to go with the trusty `for` loop for now.

Reply:
[Remove deep nesting of rubbish functional iterators](https://github.com/blockprotocol/blockprotocol/pull/928/commits/d9e4be57740420e67d18fcf2e33a405d3fa8eafc)

Reply:
I think [getEntities](https://github.com/blockprotocol/blockprotocol/blob/d9e4be57740420e67d18fcf2e33a405d3fa8eafc/libs/%40blockprotocol/graph/src/stdlib/subgraph/element/entity.ts#L23) is also a good candidate for optimisation, but we can leave it to follow up if performance is actually an issue